test-harness/src/lib.rs

```diff
@@ -42,13 +45,47 @@ pub fn _print(args: fmt::Arguments) {
     }
 }
 
+#[macro_export]
+macro_rules! runtime_handlers {
+    () => {
+        use caliptra_cpu::TrapRecord;
+
+        #[no_mangle]
+        #[inline(never)]
+        extern "C" fn exception_handler(trap_record: &TrapRecord) {
+            println!(
+                "TEST EXCEPTION mcause=0x{:08X} mscause=0x{:08X} mepc=0x{:08X}",
+                trap_record.mcause, trap_record.mscause, trap_record.mepc
+            );
+
+            // Signal non-fatal error to SOC
+            caliptra_drivers::report_fw_error_non_fatal(0xdead0);
+
+            assert!(false);
+        }
+
+        #[no_mangle]
+        #[inline(never)]
+        extern "C" fn nmi_handler(trap_record: &TrapRecord) {
+            println!(
+                "TEST NMI mcause=0x{:08X} mscause=0x{:08X} mepc=0x{:08X}",
+                trap_record.mcause, trap_record.mscause, trap_record.mepc
+            );
+
+            assert!(false);
+        }
+    };
+}
+
 #[macro_export]
 macro_rules! test_suite {
     ($($test_case: ident,)*) => {
         use core::arch::global_asm;
         use core::panic::PanicInfo;
-        use $crate::harness::Testable;
+        use caliptra_test_harness::{println, Testable};
 
+
+        #[cfg(feature = "rom")]
         global_asm!(include_str!("start.S"));
```

Comment:
Where is this file coming from?

Reply:
https://github.com/chipsalliance/caliptra-sw/blob/main/drivers/test-fw/src/bin/start.S

But this is only for ROM. RW tests use https://github.com/chipsalliance/caliptra-sw/tree/main/cpu/src for early boot code.

Open to reorganizing this though.

Reply:
Moved start.S to be part of the test-harness crate